spylon_kernel/scala_interpreter.py

```diff
@@ -300,86 +324,56 @@ def register_stderr_handler(self, handler):
 
-    def _initialize_stdout_err(self, tempdir):
-        """Redirects stdout/stderr in the Scala interpreter to two files in the
-        given tempdir and begins async tasks to poll those files for lines of text.
-
-        Parameters
-        ----------
-        tempdir : str
-            Temporary directory
-        """
-        stdout_file = os.path.abspath(os.path.join(tempdir, 'stdout'))
-        stderr_file = os.path.abspath(os.path.join(tempdir, 'stderr'))
-
-        code = 'Console.set{pipe}(new PrintStream(new FileOutputStream(new File(new java.net.URI("{filename}")), true)))'
-        code = '\n'.join([
-            'import java.io.{PrintStream, FileOutputStream, File}',
-            'import scala.Console',
-            # Set console out and error for the main thread
-            code.format(pipe="Out", filename=pathlib.Path(stdout_file).as_uri()),
-            code.format(pipe="Err", filename=pathlib.Path(stderr_file).as_uri()),
-        ])
-        self.interpret(code)
-
-        self.loop.create_task(self._poll_file(stdout_file, self.handle_stdout))
-        self.loop.create_task(self._poll_file(stderr_file, self.handle_stderr))
-
     def handle_stdout(self, line):
-        """Passes a line of Scala stdout to registered handlers.
+        """Passes a chunk of Scala stdout to registered handlers.
 
         Parameters
         ----------
         line : str
-            Line of text
+            Chunk of text
         """
         for handler in self._stdout_handlers:
             handler(line)
 
     def handle_stderr(self, line):
-        """Passes a line of Scala stderr to registered handlers.
+        """Passes a chunk of Scala stderr to registered handlers.
 
         Parameters
         ----------
         line : str
-            Line of text
+            Chunk of text
         """
         for handler in self._stderr_handlers:
             handler(line)
 
-    async def _poll_file(self, filename, fn):
-        """Busy-polls a file for lines of text and passes them to
-        the provided callback function when available.
+    def _read_stream(self, fd, fn):
+        """Reads bytes from a file descriptor, utf-8 decodes them, and passes them
+        to the provided callback function on the next IOLoop tick.
+
+        Assumes fd.read will block and should be used in a thread.
 
         Parameters
         ----------
-        filename : str
-            Filename to poll for lines of text
+        fd : file
+            File descriptor to read
         fn : callable(str) -> None
-            Callback function that handles lines of text
+            Callback function that handles chunks of text
         """
-        fd = open(filename, 'r')
         while True:
-            chars = fd.read(4096)
-            if chars:
-                fn(chars)
-                await asyncio.sleep(0, loop=self.loop)
-            else:
-                await asyncio.sleep(0.01, loop=self.loop)
+            buff = fd.read(8192)
```

Comment:
Greater-than-zero to specify that we don't want to wait until the pipe closes. Less than 65k which is (probably) the maximum pipe size these days (https://unix.stackexchange.com/questions/11946/how-big-is-the-pipe-buffer).